src/merkle.rs

```diff
@@ -41,33 +41,116 @@ pub const BUILD_DATA_BLOCK_SIZE: usize = 64 * BUILD_CHUNK_NODES;
+/// number of the above sub-tree structures (not pictured).
 ///
-/// TODO: Ord
-#[derive(Debug, Clone, Eq, PartialEq, Default)]
-pub struct MerkleTree<T, A, K, U = U2>
+/// BaseTreeArity is the arity of the base layer trees [bottom].
+/// SubTreeArity is the arity of the sub-tree layer of trees [middle].
+/// TopTreeArity is the arity of the top layer of trees [top].
+///
+/// With N and R defaulting to 0, the tree performs as a single base
+/// layer merkle tree without layers (i.e. a conventional merkle
+/// tree).
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[allow(clippy::enum_variant_names)]
+enum Data<E: Element, A: Algorithm<E>, S: Store<E>, BaseTreeArity: Unsigned, SubTreeArity: Unsigned>
+{
+    /// A BaseTree contains a single Store.
+    BaseTree(S),
+
+    /// A SubTree contains a list of BaseTrees.
+    SubTree(Vec<MerkleTree<E, A, S, BaseTreeArity>>),
+
+    /// A TopTree contains a list of SubTrees.
+    TopTree(Vec<MerkleTree<E, A, S, BaseTreeArity, SubTreeArity>>),
+}
+
+impl<E: Element, A: Algorithm<E>, S: Store<E>, BaseTreeArity: Unsigned, SubTreeArity: Unsigned>
+    Data<E, A, S, BaseTreeArity, SubTreeArity>
+{
+    /// Read-only access to the BaseTree store.
+    fn store(&self) -> Option<&S> {
+        match self {
+            Data::BaseTree(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Mutable access to the BaseTree store.
+    fn store_mut(&mut self) -> Option<&mut S> {
+        match self {
+            Data::BaseTree(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Access to the list of SubTrees.
+    #[allow(clippy::type_complexity)]
+    fn sub_trees(&self) -> Option<&Vec<MerkleTree<E, A, S, BaseTreeArity, SubTreeArity>>> {
+        match self {
+            Data::TopTree(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    // Access to the list of BaseTrees.
+    fn base_trees(&self) -> Option<&Vec<MerkleTree<E, A, S, BaseTreeArity>>> {
+        match self {
+            Data::SubTree(s) => Some(s),
+            _ => None,
+        }
+    }
+}
+
+#[allow(clippy::type_complexity)]
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct MerkleTree<E, A, S, BaseTreeArity = U2, SubTreeArity = U0, TopTreeArity = U0>
 where
-    T: Element,
-    A: Algorithm<T>,
-    K: Store<T>,
-    U: Unsigned,
+    E: Element,
+    A: Algorithm<E>,
+    S: Store<E>,
+    BaseTreeArity: Unsigned,
+    SubTreeArity: Unsigned,
+    TopTreeArity: Unsigned,
 {
-    data: K,
+    data: Data<E, A, S, BaseTreeArity, SubTreeArity>,
     leafs: usize,
+    len: usize,
     height: usize,
 
     // Cache with the `root` of the tree built from `data`. This allows to
     // not access the `Store` (e.g., access to disks in `DiskStore`).
-    root: T,
+    root: E,
 
-    _u: PhantomData<U>,
     _a: PhantomData<A>,
-    _t: PhantomData<T>,
+    _e: PhantomData<E>,
+    _bta: PhantomData<BaseTreeArity>,
+    _sta: PhantomData<SubTreeArity>,
+    _tta: PhantomData<TopTreeArity>,
```

Comment:
Not sure, but you might not need the phantomdata variants if you pass them to `Data`

Reply:
Removing them causes errors as unused.

Reply:
too bad, I guess it is the recursion